src/Limiters/limiter.jl

```diff
@@ -0,0 +1,116 @@
+
+        # This should never happen, but if it does, don't limit
+        if ρstar_e_mass[e] <= 0
+            return Qstar
+        end
+
+        # Relax constraints to ensure limiter has a solution:
+        # This is only needed if runnign with the SSP CFL>1 or
+        # due to roundoff errors.
+        if Qstar_e_mass[e] < minQ[e] * ρstar_e_mass[e]
+            minQ[e] = Qstar_e_mass[e] / ρstar_e_mass[e]
+        end
+
+        if Qstar_e_mass[e] > maxQ[e] * ρstar_e_mass[e]
+            maxQ[e] = Qstar_e_mass[e] / ρstar_e_mass[e]
+        end
+
+        local_geometry_slab = Fields.slab(space.local_geometry, e)
+
+        for iter in 1:(Nq * Nq - 1)
+            mass_change = 0.0
+            # Iterate over quadrature points
+            for i in 1:Nq, j in 1:Nq
+                wJ[i, j] = local_geometry_slab[i, j].WJ
+                ρstar_wJ[i, j] = wJ[i, j] * parent(ρstar)[i, j, 1, e]
+                # Compute the error tolerance and project qstar into the
+                # upper and lower bounds
+                if qstar[i, j, 1, e] > maxQ[e]
+                    mass_change +=
+                        (qstar[i, j, 1, e] - maxQ[e]) * ρstar_wJ[i, j]
+                    qstar[i, j, 1, e] = maxQ[e]
+                elseif qstar[i, j, 1, e] < minQ[e]
+                    mass_change -=
+                        (minQ[e] - qstar[i, j, 1, e]) * ρstar_wJ[i, j]
+                    qstar[i, j, 1, e] = minQ[e]
+                end
+            end
+
+            # By this projection of qstar to the upper/lower bounds, we
+            # have introduced a mass_change. Check if we are within the chosen
+            # tolerance
+            if abs(mass_change) <= tol * abs(Qstar_e_mass[e])
+                break
+            end
+
+            weights_sum = 0.0
+
+            # If the change was positive, the removed mass is added
+            if mass_change > 0
+                # Iterate over quadrature points
+                for i in 1:Nq, j in 1:Nq
+                    if qstar[i, j, 1, e] < maxQ[e]
+                        weights_sum += ρstar_wJ[i, j]
+                    end
+                end
+                for i in 1:Nq, j in 1:Nq
+                    if qstar[i, j, 1, e] < maxQ[e]
+                        qstar[i, j, 1, e] += mass_change / weights_sum
+                    end
+                end
+                # If the change was negative, the added mass is removed
+            else
+                # Iterate over quadrature points
+                for i in 1:Nq, j in 1:Nq
+                    if qstar[i, j, 1, e] > minQ[e]
+                        weights_sum += ρstar_wJ[i, j]
+                    end
+                end
+                for i in 1:Nq, j in 1:Nq
+                    if qstar[i, j, 1, e] > minQ[e]
+                        qstar[i, j, 1, e] += mass_change / weights_sum
+                    end
+                end
+            end
+        end
+    end
+
+    Qstar = qstar .* parent(ρstar)
```

Comment:
This is assigning a new array to `Qstar`, rather than updating the existing `Qstar` field